snakemake_storage_plugin_xrootd/__init__.py

```diff
@@ -16,178 +21,170 @@
 from snakemake_interface_storage_plugins.storage_object import (
     StorageObjectRead,
     StorageObjectWrite,
-    StorageObjectGlob,
     retry_decorator,
 )
 from snakemake_interface_storage_plugins.io import IOCacheStorageInterface
 
 
-def parse_query_params(query_params: List[str]):
-    """Parse query params from command line."""
-    return dict(param.split("=") for param in query_params)
+class XRootDFatalException(Exception):
+    """
+    Used to prevent retries for certain XRootD errors with the retry decorator
+    """
+
+    pass
 
 
-def unparse_query_params(query_params: Mapping[str, str]):
-    """Unparse query params to command line."""
-    return [f"{key}={value}" for key, value in query_params.items()]
+def _raise_fatal_error(exception: Type[Exception]):
+    if isinstance(exception, XRootDFatalException):
+        get_logger().warning("Unrecoverable error, no more retries")
+        raise exception
+
+
+# TODO configurable would be nice
+xrootd_retry = retry(
+    tries=3, delay=3, backoff=2, logger=get_logger(), fail_callback=_raise_fatal_error
+)
```

Comment:
Why using a special retry decorator here and not the one provided by snakemake-interface-storage-plugin?

Reply:
This was just because when I was testing it I would sometimes find that when it failed it would keep retrying on errors that should not really (imo) be retried on like access permissions problems. I figured maybe instead I could have a list of "veto" error codes that should not be re-tried and just fail straight away (thinking things like out of memory on the host side etc.)

I ended up re-defining the retry decorator here because I figured it was maybe the easiest way to get the custom callback function in to throw the "unretry-able" errors straight away instead of doing the usual retries but this extra decorator was a bit of an afterthought so happy to change/remove it!

Reply:
I see. This is a good idea and we can keep it like this for now. Maybe we can reintegrate it with the plugin interface at some point.